lib/vrl/compiler/src/expression/op.rs

```diff
@@ -94,8 +94,13 @@ impl Expression for Op {
             // null || ...
             Or if lhs_kind.is_null() => rhs_def,
 
-            // "foo" || ...
-            Or if !lhs_kind.is_boolean() => lhs_def,
+            Or if !lhs_kind.contains(K::Null) => lhs_def,
+
+            Or if !lhs_kind.is_boolean() => {
+                // We can remove Null from the lhs since we know that if the lhs is Null
+                // we will be taking the rhs and only the rhs type_def will then be relevant.
+                (lhs_def - K::Null).merge(rhs_def)
+            }
```

Comment:
Can we add the relevant `... || ...` comments back, similar to all the other cases?